kqueen/engines/openstack_kubespray.py

```diff
@@ -807,7 +813,40 @@ def _get_userdata(self):
         }
         return "#cloud-config\n" + yaml.dump(userdata)
 
-    def _boot_servers(self, *, name, servers_range, image, flavor, network,
+    def _set_up_security_groups(self):
+        master_sg = self.c.get_security_group(MASTER_SECURITY_GR)
+        if not master_sg:
+            master_sg = self.c.create_security_group(name=MASTER_SECURITY_GR,
+                                                     description="Security group for k8s master nodes")
+            # etcd server client API
+            self.c.create_security_group_rule(master_sg.id, protocol="tcp",
+                                              port_range_min="2379",
+                                              port_range_max="2380")
+            # k8s API
+            self.c.create_security_group_rule(master_sg.id, protocol="tcp",
+                                              port_range_min="6443",
+                                              port_range_max="6443")
+            # Calico
+            self.c.create_security_group_rule(master_sg.id, protocol="tcp",
+                                              port_range_min="179",
+                                              port_range_max="179")
+
+        common_sg = self.c.get_security_group(COMMON_SECURITY_GR)
+        if not common_sg:
+            common_sg = self.c.create_security_group(name=COMMON_SECURITY_GR,
+                                                     description="Security group for k8s slave nodes")
+
+            # Kubelet API
+            self.c.create_security_group_rule(common_sg.id, protocol="tcp",
+                                              port_range_min="10250",
+                                              port_range_max="10255")
+            # NodePort Services
+            self.c.create_security_group_rule(common_sg.id, protocol="tcp",
+                                              port_range_min="30000",
+                                              port_range_max="32767")
+        return master_sg, common_sg
```

Comment:
there is no 'deprovision' step for SC

Reply:
you will not be able to delete sg if other cluster are using it. I think it will increase deletion time witch is so big now, but sg is not occupied space so fr me it is ok to leave it